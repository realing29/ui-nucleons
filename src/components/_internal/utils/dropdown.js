import { DEFAULTS, HEIGHTS } from '../../text-field';

/**
 * Получив размер TextField, возвращает свойства для Dropdown для позиционирования строго под полем.
 * @param {string} size Размер TextField.
 * @return {Object} Свойства для Dropdown.
 */
export const placeDropdown = (size = DEFAULTS.size) => ({
  style: {
    top: `${HEIGHTS[size]}px`,
  },
});
