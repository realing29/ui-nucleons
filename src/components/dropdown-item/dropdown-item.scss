@use '../colors';

$sizes: (s: 40px, m: 48px, l: 56px, xl: 72px);

.root {
  display: flex;
  align-items: center;
  padding: 0 24px;
  &:hover {
    cursor: pointer;
    background: colors.$gray4;
  }
  @each $key, $size in $sizes {
    &.size-#{$key} {
      height: $size;
    }
  }
}
