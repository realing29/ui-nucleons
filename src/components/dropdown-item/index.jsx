import React from 'react';
import classnames from 'classnames/bind';
import styles from './dropdown-item.scss';
import PropTypes from 'prop-types';

const cx = classnames.bind(styles);

/**
 * Компонент элемента выпадающего списка.
 * @param {Object} props Свойства.
 * @param {'s'|'m'|'l'|'xl'} [props.size='m'] Размер.
 * @param {*} props.children Содержимое.
 * @param {string} [props.className] Класс.
 * @param {boolean} props.disabled Отключен ли элемент.
 * @param {boolean} props.noHover Нужно ли отключить эффект при наведении.
 * @return {ReactElement} Элемент выпадающего списка.
 */
export const DropdownItem = ({
  size = 'm',
  children,
  className,
  disabled,
  noHover,
  ...restProps
}) => (
  <div
    {...restProps}
    className={cx(
      'root',
      `size-${size}`,
      disabled && 'disabled',
      noHover && 'no-hover',
      className
    )}
  >
    {children}
  </div>
);

DropdownItem.propTypes = {
  size: PropTypes.oneOf(['s', 'm', 'l', 'xl']),
  children: PropTypes.node,
  className: PropTypes.string,
  disabled: PropTypes.bool,
  noHover: PropTypes.bool,
};
