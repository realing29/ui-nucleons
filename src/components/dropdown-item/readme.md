Элемент выпадающего списка.

### Назначение

Используется в выпадающих списках, например под полями.

### Размер и расположение

**Ширина** зависит от места использования, обычно растягивается на весь родительский блок.

**Высота** бывает 3 типов:

- s - 40px;
- m - 48px;
- l - 56px;

### Описание работы

При наведении окрашивается серым цветом (tokens:gray4)
