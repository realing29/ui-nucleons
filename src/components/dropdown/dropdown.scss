.root {
  padding: 8px 0;
  max-height: var(--max-height, 216px);
  overflow: auto;
  background: #fff;
  outline: 0;
}

// @todo сделать полосу прокрутки по макетам когда они появятся
.invisible-scroll {
  scrollbar-width: none; // Firefox
  overflow: -moz-scrollbars-none; // old Firefox
  -ms-overflow-style: none; // IE, old Edge
  &::-webkit-scrollbar {
    display: none; // Chrome, Safari, Opera
  }
}
