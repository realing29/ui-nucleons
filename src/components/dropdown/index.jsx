import React, { forwardRef } from 'react';
import { BoxShadow } from '../styling/shadows';
import styles from './dropdown.scss';
import classnames from 'classnames/bind';
import { MediumRounds } from '../styling/shapes';
import PropTypes from 'prop-types';

const cx = classnames.bind(styles);

/**
 * Компонент выпадающего контента.
 * @param {*} props.children Содержимое.
 * @param {boolean} [props.withShadow=true] Нужна ли тень.
 * @param {string} [props.className] Класс.
 * @return {ReactElement} Компонент.
 */
export const Dropdown = forwardRef(function Dropdown ({
  children,
  withShadow = true,
  className,
  ...restProps
}, ref) {
  return (
    <div
      {...restProps}
      ref={ref}
      className={cx(
        'root',
        'invisible-scroll',
        withShadow && BoxShadow.z4,
        MediumRounds.all,
        className
      )}
    >
      {children}
    </div>
  );
});

Dropdown.propTypes = {
  /**
   * Содержимое.
   */
  children: PropTypes.node,

  /**
   * Нужна ли тень.
   */
  withShadow: PropTypes.bool,

  /**
   * Класс.
   */
  className: PropTypes.string,
};
