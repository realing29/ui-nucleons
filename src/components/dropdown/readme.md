Компонент выпадающего контента. Помимо элементов списка может содержать иные элементы.

### Назначение

Выпадающие списки могут появиться под полями или кнопками.

### Размер и расположение

**Ширина** зависит от места использования.

**Высота** зависит от места использования.

### Описание работы

Имеет максимальную высоту (зависит от места использования, по умолчанию 216px).
Если содержимое имеет большую высоту, появляется прокрутка, полоса прокрутки не появляется.
