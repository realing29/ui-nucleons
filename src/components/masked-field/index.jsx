import React, { forwardRef, useEffect, useImperativeHandle, useRef, useState } from 'react';
import TextField from '../text-field';
import { InputMask } from '@krutoo/input-mask/dist/dom';
import PropTypes from 'prop-types';

const maskCommons = { placeholder: '_', pattern: /\d/ };

/**
 * Компонент текстового поля (TextField) с маской.
 * @param {Object} props Свойства.
 * @param {string} props.mask Маска.
 * @param {string} [props.value] Значение.
 * @param {Function} props.onBlur Сработает при событии blur.
 * @return {ReactElement} Компонент.
 */
export const MaskedField = forwardRef(function MaskedField ({
  mask,
  value = '',
  onBlur,
  ...restOptions
}, ref) {
  const [inputMask, setInputMask] = useState();
  const [inputState, setInputState] = useState({ value, cleanValue: '' });
  const innerRef = useRef();

  useImperativeHandle(ref, () => innerRef.current);

  useEffect(() => {
    const im = InputMask(innerRef.current, {
      ...maskCommons,
      mask,
      onChange: setInputState,
    });

    im.setValue(value);
    setInputMask(im);

    return () => im.disable();
  }, [mask]);

  useEffect(() => {
    inputMask && inputMask.setValue(value);
  }, [value, inputMask]);

  return (
    <TextField
      {...restOptions}
      ref={innerRef}
      restPlaceholder={{
        value: mask.slice(inputState.value.length),
        shiftValue: inputState.value,
      }}

      // получаем данные маски только по событию "blur"
      onBlur={e => onBlur && onBlur(e, inputState)}
    />
  );
});

MaskedField.propTypes = {
  /**
   * Маска.
   */
  mask: PropTypes.string.isRequired,

  /**
   * Значение.
   */
  value: PropTypes.string,

  /**
   * Сработает при событии blur.
   */
  onBlur: PropTypes.func,
};
