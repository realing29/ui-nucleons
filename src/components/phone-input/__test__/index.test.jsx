import React from 'react';
import { mount } from 'enzyme';
import { act } from 'react-dom/test-utils';
import { PhoneInput } from '..';
import TextField from '../../text-field';
import { Dropdown } from '../../dropdown';
import { DropdownItem } from '../../dropdown-item';
import { MaskedField } from '../../masked-field';
import { countries } from '../presets';

describe('<PhoneInput />', () => {
  it('should render without props', () => {
    const wrapper = mount(<PhoneInput />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render correct countries in dropdown', () => {
    const wrapper = mount(<PhoneInput />);
    expect(wrapper.find(Dropdown)).toHaveLength(0);

    act(() => {
      wrapper.find('[data-testid="phone-input:dropdown-opener"]').simulate('click');
    });
    wrapper.update();
    expect(wrapper.find(Dropdown)).toHaveLength(1);
  });

  it('should call onCountrySelect on change country', () => {
    const onCountrySelect = jest.fn();
    const wrapper = mount(
      <PhoneInput
        onCountrySelect={onCountrySelect}
      />
    );

    expect(onCountrySelect).not.toHaveBeenCalled();
    wrapper.find('[data-testid="phone-input:dropdown-opener"]').simulate('click');
    expect(wrapper.find(Dropdown)).toHaveLength(1);

    act(() => {
      wrapper.find(DropdownItem).at(0).prop('onClick')();
    });
    wrapper.update();

    expect(wrapper.find(Dropdown)).toHaveLength(0);
    expect(onCountrySelect).toHaveBeenCalledTimes(1);
  });

  it('should open and close countries popup properly', () => {
    const wrapper = mount(<PhoneInput />);

    expect(wrapper.find(Dropdown)).toHaveLength(0);

    act(() => {
      wrapper.find('[data-testid="phone-input:dropdown-opener"]').simulate('click');
    });
    wrapper.update();
    expect(wrapper.find(Dropdown)).toHaveLength(1);

    act(() => {
      document.documentElement.dispatchEvent(new MouseEvent('click'));
    });
    wrapper.update();
    expect(wrapper.find(Dropdown)).toHaveLength(0);
  });

  it('should pass mask state to onBlur callback', () => {
    const spy = jest.fn();

    const wrapper = mount(
      <PhoneInput onBlur={spy} />
    );

    expect(spy).toHaveBeenCalledTimes(0);

    act(() => {
      wrapper.find(TextField).find('input').simulate('blur');
    });

    expect(spy).toHaveBeenCalledTimes(1);
    expect(spy.mock.calls[0][0].type).toBe('blur');
    expect(spy.mock.calls[0][1].cleanValue).toBe('7');
  });

  it('should define initial mask properly', () => {
    const wrapper = mount(
      <PhoneInput value='375112223344' />
    );

    expect(wrapper.find(MaskedField).prop('mask')).toBe(countries.belarus.mask);
  });

  it('should handle "value" prop change', () => {
    const spy = jest.fn();

    const wrapper = mount(
      <PhoneInput value='375112223344' onBlur={spy} />
    );

    act(() => {
      wrapper.find(TextField).find('input').simulate('blur');
    });

    expect(spy).toHaveBeenCalledTimes(1);
    expect(spy.mock.calls[0][0].type).toBe('blur');
    expect(spy.mock.calls[0][1].cleanValue).toBe('375112223344');

    wrapper.setProps({ value: '375000000000' });

    act(() => {
      wrapper.find(TextField).find('input').simulate('blur');
    });

    expect(spy).toHaveBeenCalledTimes(2);
    expect(spy.mock.calls[1][0].type).toBe('blur');
    expect(spy.mock.calls[1][1].cleanValue).toBe('375000000000');
  });
});
