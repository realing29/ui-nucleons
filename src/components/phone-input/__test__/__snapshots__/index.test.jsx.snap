// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PhoneInput /> Other: should render just text field with able to input only digits 1`] = `
<PhoneInput
  onBlur={[MockFunction]}
  onChange={[MockFunction]}
  value="0000000"
>
  <div
    className="root"
  >
    <ForwardRef(TextField)
      className="field"
      endAdornment={
        <div
          className="adornment"
          data-testid="phone-input:dropdown-opener"
          onClick={[Function]}
        >
          <img
            alt=""
            height={24}
            src={Object {}}
            width={24}
          />
          <DownSVG
            className="M-l__1"
            fill="#9e9e9e"
          />
        </div>
      }
      label="Телефон"
      multiline={false}
      onBlur={[Function]}
      onChange={[Function]}
    >
      <div
        className="root field"
      >
        <div
          className="reset input-block variant-desktop size-l with-label rounds-sm__all"
          data-testid="text-field:block"
          onClick={[Function]}
        >
          <div
            className="main-column"
          >
            <label
              className="label variant-desktop as-placeholder"
            >
              Телефон
            </label>
            <div
              className="input-wrapper"
            >
              <ForwardRef(BaseInput)
                className="base-input variant-desktop size-l"
                data-testid="text-field:field"
                multiline={false}
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
              >
                <div
                  className="reset root base-input variant-desktop size-l"
                >
                  <input
                    className="field reset"
                    data-testid="text-field:field"
                    onBlur={[Function]}
                    onChange={[Function]}
                    onFocus={[Function]}
                    onInput={[Function]}
                  />
                </div>
              </ForwardRef(BaseInput)>
            </div>
          </div>
          <Box
            alignItems="center"
            display="flex"
            marginLeft={4}
          >
            <div
              className="display-flex align-items-center M-l__4"
            >
              <div
                className="adornment"
                data-testid="phone-input:dropdown-opener"
                onClick={[Function]}
              >
                <img
                  alt=""
                  height={24}
                  src={Object {}}
                  width={24}
                />
                <DownSVG
                  className="M-l__1"
                  fill="#9e9e9e"
                >
                  <svg
                    className="M-l__1"
                    fill="#9e9e9e"
                    height={16}
                    viewBox="0 0 16 16"
                    width={16}
                  >
                    <path
                      clipRule="evenodd"
                      d="M3.247 6.342a1 1 0 011.412-.095L8 9.171l3.341-2.924a1 1 0 011.318 1.506l-4 3.5a1 1 0 01-1.317 0l-4-3.5a1 1 0 01-.095-1.411z"
                      fillRule="evenodd"
                    />
                  </svg>
                </DownSVG>
              </div>
            </div>
          </Box>
        </div>
      </div>
    </ForwardRef(TextField)>
  </div>
</PhoneInput>
`;

exports[`<PhoneInput /> should render without props 1`] = `
<PhoneInput>
  <div
    className="root"
  >
    <ForwardRef(MaskedField)
      className="field"
      endAdornment={
        <div
          className="adornment"
          data-testid="phone-input:dropdown-opener"
          onClick={[Function]}
        >
          <img
            alt=""
            height={24}
            src={Object {}}
            width={24}
          />
          <DownSVG
            className="M-l__1"
            fill="#9e9e9e"
          />
        </div>
      }
      label="Телефон"
      mask="+7 (___) ___-__-__"
      multiline={false}
      onBlur={[Function]}
      value=""
    >
      <ForwardRef(TextField)
        className="field"
        endAdornment={
          <div
            className="adornment"
            data-testid="phone-input:dropdown-opener"
            onClick={[Function]}
          >
            <img
              alt=""
              height={24}
              src={Object {}}
              width={24}
            />
            <DownSVG
              className="M-l__1"
              fill="#9e9e9e"
            />
          </div>
        }
        label="Телефон"
        multiline={false}
        onBlur={[Function]}
        restPlaceholder={
          Object {
            "shiftValue": "+7 (",
            "value": "___) ___-__-__",
          }
        }
      >
        <div
          className="root field"
        >
          <div
            className="reset input-block variant-desktop size-l with-label rounds-sm__all"
            data-testid="text-field:block"
            onClick={[Function]}
          >
            <div
              className="main-column"
            >
              <label
                className="label variant-desktop filled"
              >
                Телефон
              </label>
              <div
                className="input-wrapper"
              >
                <ForwardRef(BaseInput)
                  className="base-input variant-desktop size-l"
                  data-testid="text-field:field"
                  multiline={false}
                  onBlur={[Function]}
                  onFocus={[Function]}
                  restPlaceholder={
                    Object {
                      "shiftValue": "+7 (",
                      "value": "___) ___-__-__",
                    }
                  }
                >
                  <div
                    className="reset root base-input variant-desktop size-l"
                  >
                    <span
                      aria-hidden={true}
                      className="fake-text"
                    >
                      <span
                        className="invisible-value"
                      >
                        +7 (
                      </span>
                      <span
                        className="placeholder"
                      >
                        ___) ___-__-__
                      </span>
                    </span>
                    <input
                      className="field reset"
                      data-testid="text-field:field"
                      onBlur={[Function]}
                      onFocus={[Function]}
                      onInput={[Function]}
                    />
                  </div>
                </ForwardRef(BaseInput)>
              </div>
            </div>
            <Box
              alignItems="center"
              display="flex"
              marginLeft={4}
            >
              <div
                className="display-flex align-items-center M-l__4"
              >
                <div
                  className="adornment"
                  data-testid="phone-input:dropdown-opener"
                  onClick={[Function]}
                >
                  <img
                    alt=""
                    height={24}
                    src={Object {}}
                    width={24}
                  />
                  <DownSVG
                    className="M-l__1"
                    fill="#9e9e9e"
                  >
                    <svg
                      className="M-l__1"
                      fill="#9e9e9e"
                      height={16}
                      viewBox="0 0 16 16"
                      width={16}
                    >
                      <path
                        clipRule="evenodd"
                        d="M3.247 6.342a1 1 0 011.412-.095L8 9.171l3.341-2.924a1 1 0 011.318 1.506l-4 3.5a1 1 0 01-1.317 0l-4-3.5a1 1 0 01-.095-1.411z"
                        fillRule="evenodd"
                      />
                    </svg>
                  </DownSVG>
                </div>
              </div>
            </Box>
          </div>
        </div>
      </ForwardRef(TextField)>
    </ForwardRef(MaskedField)>
  </div>
</PhoneInput>
`;
